File: Ubuntu18.04/Samples/Base/DCAM650/MultiConnection/MultiConnection.cpp
#include <thread>
#include <iostream>
#include "VzenseNebula_api.h"
#define frameSpace 10
using namespace std;

int main() {
	cout << "---MultiConnection---"<< endl;

	//about dev
	uint32_t deviceCount = 0;
	VzReturnStatus status = VzRetOthers;
	VzDeviceHandle *deviceHandle = nullptr;
	VzDeviceInfo* pDeviceListInfo = nullptr;

	
	//about Frame
	
	VzFrameReady FrameReady = { 0 };
	VzFrame Depth = { 0 };

	//SDK Initialize
	status = VZ_Initialize();
	if (status != VzReturnStatus::VzRetOK)
	{
		cout << "VzInitialize failed status:" <<status << endl;
		system("pause");
		return -1;
	}

	//1.Search and notice the count of devices.
	//2.get infomation of the devices. 
	//3.open devices accroding to the info.
GET:
	status = VZ_GetDeviceCount(&deviceCount);
	if (status != VzReturnStatus::VzRetOK)
	{
		cout << "VzGetDeviceCount failed! make sure the DCAM is connected" << endl;
		this_thread::sleep_for(chrono::seconds(1));
		goto GET;
	}
	cout << "Get device count: " << deviceCount << endl;
	if (deviceCount < 2)
	{
		this_thread::sleep_for(chrono::seconds(1));
		goto GET;
	}

	deviceHandle = new VzDeviceHandle[deviceCount];
	pDeviceListInfo = new VzDeviceInfo[deviceCount];

	status = VZ_GetDeviceInfoList(deviceCount, pDeviceListInfo);
	if (status != VzReturnStatus::VzRetOK)
	{
		cout << "GetDeviceListInfo failed status:" << status << endl;
		return -1;
	}
	else
	{
		if (VzConnected != pDeviceListInfo[0].status)
		{
			cout << "connect statu" << pDeviceListInfo[0].status << endl;
			cout << "Call VZ_OpenDevice with connect status :" << VzConnected << endl;
			return -1;
		}
	}
	for (int i = 0; i < deviceCount; i++)
	{
		cout << "CameraIndex: " << i << endl;
		cout << "uri:" << pDeviceListInfo[i].uri << endl
			<< "alias:" << pDeviceListInfo[i].alias << endl
			<< "connectStatus:" << pDeviceListInfo[i].status << endl;
	}
	for(int i = 0;i < deviceCount; i++)
	{
		status = VZ_OpenDeviceByUri(pDeviceListInfo[i].uri, &deviceHandle[i]);
		if (status != VzReturnStatus::VzRetOK)
		{
			cout << "OpenDevice failed status:" <<status << endl;
			return -1;
		}
	}

    for (int i = 0; i < deviceCount; i++)
    {
        //Starts capturing the image stream
        status = VZ_StartStream(deviceHandle[i]);
        if (status != VzReturnStatus::VzRetOK)
        {
            cout << "VZ_StartStream failed status:" << status << endl;
            return -1;
        }
    }

    //Wait for the device to upload image data
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));

	//1.ReadNextFrame.
	//2.Get depth Frame acoording to Ready flag.
	for (int j = 0; j < frameSpace; j++)
	{
		for (int i = 0; i < deviceCount; i++)
		{
			if (deviceHandle[i])
			{
				VzFrame depthFrame = { 0 };
				VzFrameReady frameReady = { 0 };
				status = VZ_GetFrameReady(deviceHandle[i], 1200, &frameReady);

				if (status != VzRetOK)
				{
					cout << pDeviceListInfo[i].alias <<"  VZ_GetFrameReady failed status:" << status << endl;
					continue;
				}

				//Get depth frame, depth frame only output in following data mode
				if (1 == frameReady.depth)
				{
					status = VZ_GetFrame(deviceHandle[i], VzDepthFrame, &depthFrame);

					if (status == VzRetOK && depthFrame.pFrameData != NULL)
					{
						cout << pDeviceListInfo[i].alias << " frameIndex :" << depthFrame.frameIndex << endl;
					}
					else
					{
						cout << pDeviceListInfo[i].alias << "VZ_GetFrame VzDepthFrame status:" << status  << endl;
					}
				}
			}

		}
	}
 
	//1.close device
	//2.SDK shutdown
	for (int i = 0; i < deviceCount; i++)
	{
		status = VZ_StopStream(deviceHandle[i]);
		if (status != VzReturnStatus::VzRetOK)
		{
			cout << "VZ_StopStream failed status:" <<status<< endl;
		}
		status = VZ_CloseDevice(&deviceHandle[i]);
		if (status != VzReturnStatus::VzRetOK)
		{
			cout << "VZ_CloseDevice failed status:" <<status<< endl;
		}
	}
	status = VZ_Shutdown();
	if (status != VzReturnStatus::VzRetOK)
	{
		cout << "VZ_Shutdown failed status:" <<status<< endl;
		return -1;
	}

	cout << "--end--"<< endl;

	return 0;
}

File: Python/Samples/DCAM650/MultiConnection/MultiConnection.py
from pickle import FALSE, TRUE
import sys
sys.path.append('../../../')

from API.VzenseDS_api import *
import time

camera = VzenseTofCam()


camera_count = camera.VZ_GetDeviceCount()
retry_count = 100
while camera_count < 2 and retry_count > 0:
    retry_count = retry_count-1
    camera_count = camera.VZ_GetDeviceCount()
    time.sleep(1)
    print("scaning......   ",retry_count)


if camera_count < 2: 
    print("there are no camera or only one camera")
    exit()

print("cam count :",camera_count)
cameras = []

ret, device_infolist=camera.VZ_GetDeviceInfoList(camera_count)
if ret==0:
    for i in range(camera_count): 
        print('cam uri:  ' + str(device_infolist[i].uri))
        cam = VzenseTofCam()
        ret = cam.VZ_OpenDeviceByUri(device_infolist[i].uri)
        if  ret == 0:
            print(device_infolist[i].alias,"open successful")
            cameras.append(cam)
        else:
            print(device_infolist[i].alias,'VZ_OpenDeviceByUri failed: ' + str(ret))    
else:
    print(' failed:' + ret)  
    exit()  

for i in range(camera_count): 
    ret = cameras[i].VZ_StartStream()       
    if  ret == 0:
        print(device_infolist[i].alias,"start stream successful")
    else:
        print(device_infolist[i].alias,'VZ_StartStream failed: ' + str(ret))  

# show image 

while 1:
    for i in range(camera_count): 
        ret, frameready = cameras[i].VZ_GetFrameReady(c_uint16(1200))
        if  ret !=0:
            print("VZ_GetFrameReady failed:",ret)
            continue
                        
        if  frameready.depth:      
            ret,depthframe = cameras[i].VZ_GetFrame(VzFrameType.VzDepthFrame)
            if  ret == 0:
                print(device_infolist[i].alias,"  depth frameindex: ",depthframe.frameIndex)
            else:
                print("VZ_GetFrame error", ret)
        if  frameready.ir:
            ret,irframe = cameras[i].VZ_GetFrame(VzFrameType.VzIRFrame)
            if  ret == 0:
                print(device_infolist[i].alias,"  ir frameindex: ",irframe.frameIndex)
            else:
                print("VZ_GetFrame error", ret)

for i in range(camera_count): 
    
    ret = cameras[i].VZ_StopStream()       
    if  ret == 0:
        print("stop stream successful")
    else:
        print('VZ_StopStream failed: ' + str(ret))  

    ret = cameras[i].VZ_CloseDevice()       
    if  ret == 0:
        print("close device successful")
    else:
        print('VZ_CloseDevice failed: ' + str(ret))  
    
           